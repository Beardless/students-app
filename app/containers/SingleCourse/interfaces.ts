export interface ModuleInterface {
  id: string;
  title: string;
  description: string;
  position: number;
  created: string;
  updated: string;
  course: {
    id: number;
    title: string;
  };
  lessons: { id: string; title: string }[];
}
