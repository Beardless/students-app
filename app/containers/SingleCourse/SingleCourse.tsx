import * as React from "react";
import CourseModuleItem from "../../components/CourseModuleItem";
import { match } from "react-router-dom";
import { ModuleInterface } from "./interfaces";
import LoadingIndicator from "../../components/LoadingIndicator";

interface SingleCourseProps {
  match: match<{ courseId: string }>;
  getCourse: (id: string) => void;
  error: boolean | string;
  loading: boolean;
  modules: [ModuleInterface];
}

class SingleCourse extends React.Component<SingleCourseProps> {
  state = {
    expanded: null
  };

  componentDidMount(): void {
    const { courseId } = this.props.match.params;
    this.props.getCourse(courseId);
  }

  handleChange = panel => (event, expanded) => {
    this.setState({
      expanded: expanded ? panel : false
    });
  };

  render(): React.ReactNode {
    const { error, loading, modules, match } = this.props;

    let currentCourse = null;

    if (modules) {
      const [firstModule] = modules;
      currentCourse = firstModule.course;
    }

    if (error) {
      return <div>{error}</div>;
    }
    if (loading) {
      return <LoadingIndicator />;
    }

    return (
      <section>
        <h2>{currentCourse && currentCourse.title}:</h2>
        {modules &&
          modules.map(module => {
            return (
              <CourseModuleItem
                match={match}
                item={module}
                key={module.id}
                expanded={this.state.expanded}
                onChange={this.handleChange}
              />
            );
          })}
      </section>
    );
  }
}

export default SingleCourse;
