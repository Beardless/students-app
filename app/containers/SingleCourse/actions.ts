import {
  COURSE_ACTION,
  COURSE_ACTION_SUCCESS,
  COURSE_ACTION_ERROR
} from "./constants";
import { ModuleInterface } from "./interfaces";

export function courseAction(
  id: string
): { type: string; payload: { id: string } } {
  return {
    type: COURSE_ACTION,
    payload: { id }
  };
}

export function courseActionSuccess(
  modules: [ModuleInterface]
): { type: string; modules: [ModuleInterface] } {
  return {
    type: COURSE_ACTION_SUCCESS,
    modules
  };
}

export function courseActionError(
  error: string
): { type: string; error: string } {
  return {
    type: COURSE_ACTION_ERROR,
    error
  };
}
