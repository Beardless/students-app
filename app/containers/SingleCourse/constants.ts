export const COURSE_ACTION: string = "courses/SingleCourse/COURSE_ACTION";
export const COURSE_ACTION_SUCCESS: string =
  "courses/SingleCourse/COURSE_ACTION_SUCCESS";
export const COURSE_ACTION_ERROR: string =
  "courses/SingleCourse/COURSE_ACTION_ERROR";
