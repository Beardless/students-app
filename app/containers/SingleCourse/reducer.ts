import { fromJS } from "immutable";

import {
  COURSE_ACTION,
  COURSE_ACTION_SUCCESS,
  COURSE_ACTION_ERROR
} from "./constants";

// The initial state of the App
const initialState = fromJS({
  loading: false,
  error: false,
  modules: false
});

function appReducer(state = initialState, action) {
  switch (action.type) {
    case COURSE_ACTION:
      return state.set("loading", true).set("error", false);
    case COURSE_ACTION_SUCCESS:
      return state.set("loading", false).set("modules", action.modules);
    case COURSE_ACTION_ERROR:
      return state.set("loading", false).set("error", action.error);
    default:
      return state;
  }
}

export default appReducer;
