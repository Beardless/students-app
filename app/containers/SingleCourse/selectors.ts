/**
 * Single Course selectors
 */

import { createSelector } from "reselect";

const selectSingleCourse = state => state.get("singleCourse");

const makeSelectSingleCourseLoading = () =>
  createSelector(
    selectSingleCourse,
    singleCourse => singleCourse.get("loading")
  );

const makeSelectSingleCourseError = () =>
  createSelector(
    selectSingleCourse,
    singleCourse => singleCourse.get("error")
  );

const makeSelectSingleCourseModules = () =>
  createSelector(
    selectSingleCourse,
    singleCourse => singleCourse.get("modules")
  );

export {
  selectSingleCourse,
  makeSelectSingleCourseLoading,
  makeSelectSingleCourseError,
  makeSelectSingleCourseModules
};
