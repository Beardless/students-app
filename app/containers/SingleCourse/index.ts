import { compose } from "redux";
import { connect } from "react-redux";
import SingleCourse from "./SingleCourse";
import { createStructuredSelector } from "reselect";
import { courseAction } from "./actions";
import {
  makeSelectSingleCourseModules,
  makeSelectSingleCourseError,
  makeSelectSingleCourseLoading
} from "./selectors";
import singleCourseReducer from "./reducer";
import singleCourseSaga from "./saga";
import injectSaga from "../../utils/injectSaga";
import injectReducer from "../../utils/injectReducer";

const mapStateToProps = createStructuredSelector({
  modules: makeSelectSingleCourseModules(),
  error: makeSelectSingleCourseError(),
  loading: makeSelectSingleCourseLoading()
});

const mapDispatchToProps = dispatch => ({
  getCourse: id => dispatch(courseAction(id))
});

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps
);

const withSingleCourseReducer = injectReducer({
  key: "singleCourse",
  reducer: singleCourseReducer
});
const withSingleCourseSaga = injectSaga({
  key: "singleCourse",
  saga: singleCourseSaga
});

export default compose<any>(
  withConnect,
  withSingleCourseReducer,
  withSingleCourseSaga
)(SingleCourse);

export { mapDispatchToProps };
