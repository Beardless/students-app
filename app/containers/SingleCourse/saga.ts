import { COURSE_ACTION } from "./constants";
import { call, put, takeEvery } from "redux-saga/effects";
import { getAuthorizationHeaders } from "../../utils/userUtils";
import { courseActionError, courseActionSuccess } from "./actions";

// @ts-ignore
import request from "utils/request";

export function* getCourse({ payload }) {
  try {
    const { id: courseId } = payload;
    const options = {
      method: "GET",
      url: `/courses/${courseId}/modules`,
      headers: getAuthorizationHeaders()
    };
    const { data: modules } = yield call(request, options);

    yield put(courseActionSuccess(modules));
  } catch (err) {
    const { message } = err.response.data;
    yield put(courseActionError(message));
  }
}

export default function* getSingleCourse() {
  // @ts-ignore
  yield takeEvery(COURSE_ACTION, getCourse);
}
