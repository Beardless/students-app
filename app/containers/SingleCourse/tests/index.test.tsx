import * as React from "react";
import SingleCourse from "../SingleCourse";
import { shallow } from "enzyme";
import * as course from "./mocks/course.json";
import { mapDispatchToProps } from "../index";
import { courseAction } from "../actions";
import { Location } from "history";

describe("<SingleCourse />", () => {
  it("should render correctly", () => {
    const renderedComponent = shallow(
      <SingleCourse
        location={{ pathname: "", search: "", state: "", hash: "git" }}
        match={{ params: { courseId: "1" }, isExact: true, path: "", url: "" }}
        getCourse={() => {}}
        error={false}
        loading={false}
        course={course}
      />
    );

    expect(renderedComponent).toMatchSnapshot();
  });

  describe("mapDispatchToProps", () => {
    describe("getCourse", () => {
      it("should be injected", function() {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.getCourse).toBeDefined();
      });

      it("should dispatch courseAction when called", () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        const id = "1";
        result.getCourse(id);
        expect(dispatch).toHaveBeenCalledWith(courseAction(id));
      });
    });
  });
});
