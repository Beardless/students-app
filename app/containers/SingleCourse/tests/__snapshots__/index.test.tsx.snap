// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleCourse /> should render correctly 1`] = `
<section>
  <Jumbotron
    subtitle="Description of first example course"
    title="First example course"
  />
  <div
    className="column-container"
  >
    <div
      className="modules"
    >
      <CourseModuleItem
        item={
          Object {
            "course": Object {
              "id": 1,
              "title": "Test title of example course",
            },
            "created": "2019-03-08T00:00:00+01:00",
            "description": "Test description of example module",
            "id": "fb8fbc90-2a0f-45b3-83ee-bb6e9d1bc402",
            "lessons": Array [
              Object {
                "completed": null,
                "duration_in_minutes": 60,
                "href": Object {
                  "cover_image_url": "simplestring.jpg",
                },
                "id": "53ca8a0a-b5a9-40e9-b5b7-0f9d226a4d8b",
                "title": "Test title of example lesson",
              },
            ],
            "position": 0,
            "progress": Object {
              "completed_lesson": "1",
              "completed_percentage": "10",
              "completed_time": "10",
              "total_duration": "100",
            },
            "title": "Test title of example module",
            "updated": "2019-03-11T19:41:18+01:00",
          }
        }
        key="fb8fbc90-2a0f-45b3-83ee-bb6e9d1bc402"
        match={
          Object {
            "isExact": true,
            "params": Object {
              "courseId": "1",
            },
            "path": "",
            "url": "",
          }
        }
      />
    </div>
    <div
      className="authors"
    >
      <WithStyles(Typography)
        variant="h5"
      />
      <Author
        author={
          Object {
            "bio": "Lorem Ipsum of bio",
            "href": Object {
              "picture": "examples.com/example.jpeg",
            },
            "name": "John Doe",
            "picture": "example.png",
          }
        }
        key="John Doe"
      />
    </div>
  </div>
</section>
`;
